File: tugtalk/src/tugbank-client.ts
// Direct bun:sqlite client for the tugbank defaults store.
// Replaces Bun.spawnSync(['tugbank', ...]) calls with in-process SQLite access.

import { Database } from "bun:sqlite";
import { homedir } from "node:os";
import { join } from "node:path";

// ── Value kind discriminators (must match tugbank-core/src/value.rs) ─────────

const KIND_NULL = 0;
const KIND_BOOL = 1;
const KIND_I64 = 2;
const KIND_F64 = 3;
const KIND_STRING = 4;
const KIND_BYTES = 5;
const KIND_JSON = 6;

// ── Types ─────────────────────────────────────────────────────────────────────

export type TugbankValue = boolean | number | string | object | Buffer | null;

interface EntryRow {
  value_kind: number;
  value_i64: number | null;
  value_f64: number | null;
  value_text: string | null;
  value_blob: Buffer | null;
}

interface DomainEntryRow extends EntryRow {
  key: string;
}

// ── Value decode/encode ───────────────────────────────────────────────────────

function decodeValue(row: EntryRow): TugbankValue {
  switch (row.value_kind) {
    case KIND_NULL:
      return null;
    case KIND_BOOL:
      return (row.value_i64 ?? 0) !== 0;
    case KIND_I64:
      return row.value_i64 ?? 0;
    case KIND_F64:
      return row.value_f64 ?? 0;
    case KIND_STRING:
      return row.value_text ?? "";
    case KIND_BYTES:
      return row.value_blob ?? Buffer.alloc(0);
    case KIND_JSON:
      return JSON.parse(row.value_text ?? "null") as object;
    default:
      throw new Error(`Unknown value_kind: ${row.value_kind}`);
  }
}

interface EncodedValue {
  kind: number;
  i64: number | null;
  f64: number | null;
  text: string | null;
  blob: Buffer | null;
}

function encodeValue(value: TugbankValue): EncodedValue {
  if (value === null) {
    return { kind: KIND_NULL, i64: null, f64: null, text: null, blob: null };
  }
  if (typeof value === "boolean") {
    return { kind: KIND_BOOL, i64: value ? 1 : 0, f64: null, text: null, blob: null };
  }
  if (typeof value === "number") {
    if (Number.isInteger(value)) {
      return { kind: KIND_I64, i64: value, f64: null, text: null, blob: null };
    }
    return { kind: KIND_F64, i64: null, f64: value, text: null, blob: null };
  }
  if (typeof value === "string") {
    return { kind: KIND_STRING, i64: null, f64: null, text: value, blob: null };
  }
  if (Buffer.isBuffer(value)) {
    return { kind: KIND_BYTES, i64: null, f64: null, text: null, blob: value };
  }
  // object — treat as JSON
  return { kind: KIND_JSON, i64: null, f64: null, text: JSON.stringify(value), blob: null };
}

function nowRfc3339(): string {
  return new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
}

// ── TugbankClient ─────────────────────────────────────────────────────────────

/**
 * Direct bun:sqlite client for the tugbank defaults store at ~/.tugbank.db.
 *
 * Provides get/set/readDomain/listDomains with an in-memory domain cache.
 * The cache is invalidated by polling PRAGMA data_version every 500 ms so
 * changes written by other processes (e.g. the tugbank CLI) are picked up
 * automatically.
 *
 * Call close() to release the database connection and stop the poll timer.
 */
export class TugbankClient {
  private readonly db: Database;
  private readonly domainCache = new Map<string, Record<string, TugbankValue>>();
  private lastDataVersion: number = -1;
  private readonly pollTimer: ReturnType<typeof setInterval>;

  constructor(dbPath?: string) {
    const path = dbPath ?? join(homedir(), ".tugbank.db");
    this.db = new Database(path, { create: true });

    // Match the pragmas applied by tugbank-core on every open.
    this.db.run("PRAGMA journal_mode = WAL");
    this.db.run("PRAGMA foreign_keys = ON");
    this.db.run("PRAGMA busy_timeout = 5000");
    this.db.run("PRAGMA synchronous = NORMAL");

    // Initialise our baseline data_version.
    this.lastDataVersion = this.readDataVersion();

    // Poll for external changes every 500 ms.
    this.pollTimer = setInterval(() => {
      this.checkForChanges();
    }, 500);

    // Don't let the timer prevent process exit.
    if (this.pollTimer.unref) {
      this.pollTimer.unref();
    }
  }

  // ── Public API ──────────────────────────────────────────────────────────────

  /**
   * Read a single key from a domain.
   * Returns null if the key does not exist.
   */
  get(domain: string, key: string): TugbankValue {
    const stmt = this.db.prepare<EntryRow, [string, string]>(
      `SELECT value_kind, value_i64, value_f64, value_text, value_blob
       FROM entries WHERE domain = ? AND key = ?`
    );
    const row = stmt.get(domain, key);
    if (!row) return null;
    return decodeValue(row);
  }

  /**
   * Write a single key/value pair to a domain.
   * Creates the domain row if it does not exist and bumps the generation.
   */
  set(domain: string, key: string, value: TugbankValue): void {
    const enc = encodeValue(value);
    const now = nowRfc3339();

    this.db.transaction(() => {
      // Ensure domain row exists.
      this.db.run(
        `INSERT OR IGNORE INTO domains (name, generation, updated_at) VALUES (?, 0, ?)`,
        [domain, now]
      );
      // Upsert the entry.
      this.db.run(
        `INSERT INTO entries
           (domain, key, value_kind, value_i64, value_f64, value_text, value_blob, updated_at)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?)
         ON CONFLICT(domain, key) DO UPDATE SET
           value_kind = excluded.value_kind,
           value_i64  = excluded.value_i64,
           value_f64  = excluded.value_f64,
           value_text = excluded.value_text,
           value_blob = excluded.value_blob,
           updated_at = excluded.updated_at`,
        [domain, key, enc.kind, enc.i64, enc.f64, enc.text, enc.blob, now]
      );
      // Bump generation.
      this.db.run(
        `UPDATE domains SET generation = generation + 1, updated_at = ? WHERE name = ?`,
        [now, domain]
      );
    })();

    // Invalidate cache for this domain.
    this.domainCache.delete(domain);
  }

  /**
   * Read all key/value pairs for a domain as a plain object.
   * Returns a cached snapshot; cache is refreshed when data_version changes.
   */
  readDomain(domain: string): Record<string, TugbankValue> {
    const cached = this.domainCache.get(domain);
    if (cached) return cached;

    const rows = this.db
      .prepare<DomainEntryRow, [string]>(
        `SELECT key, value_kind, value_i64, value_f64, value_text, value_blob
         FROM entries WHERE domain = ?`
      )
      .all(domain);

    const result: Record<string, TugbankValue> = {};
    for (const row of rows) {
      result[row.key] = decodeValue(row);
    }

    this.domainCache.set(domain, result);
    return result;
  }

  /**
   * List all domain names that have at least one entry.
   */
  listDomains(): string[] {
    const rows = this.db
      .prepare<{ name: string }, []>(`SELECT name FROM domains ORDER BY name`)
      .all();
    return rows.map((r) => r.name);
  }

  /**
   * Release the database connection and stop the poll timer.
   */
  close(): void {
    clearInterval(this.pollTimer);
    this.db.close();
  }

  // ── Private helpers ─────────────────────────────────────────────────────────

  private readDataVersion(): number {
    const row = this.db.prepare<{ data_version: number }, []>("PRAGMA data_version").get();
    return row?.data_version ?? 0;
  }

  private checkForChanges(): void {
    try {
      const current = this.readDataVersion();
      if (current !== this.lastDataVersion) {
        this.lastDataVersion = current;
        this.domainCache.clear();
      }
    } catch {
      // Ignore errors during poll (db may be closing).
    }
  }
}

// ── Singleton ─────────────────────────────────────────────────────────────────

/**
 * Module-level singleton TugbankClient.
 *
 * Created lazily on first access so that importing this module in test
 * environments that don't have ~/.tugbank.db doesn't immediately fail.
 */
let _client: TugbankClient | null = null;

export function getTugbankClient(): TugbankClient {
  if (!_client) {
    _client = new TugbankClient();
  }
  return _client;
}

File: tugapp/Sources/TugbankClient.swift
import Foundation
import SQLite3

// MARK: - TugbankValue

/// A typed value stored in a tugbank domain.
/// Matches the Rust `Value` enum and the SQL column encoding.
enum TugbankValue {
    case null
    case bool(Bool)
    case i64(Int64)
    case f64(Double)
    case string(String)
    case json(Any)

    // value_kind discriminators (must match tugbank-core/src/value.rs)
    static let kindNull: Int32 = 0
    static let kindBool: Int32 = 1
    static let kindI64: Int32 = 2
    static let kindF64: Int32 = 3
    static let kindString: Int32 = 4
    static let kindBytes: Int32 = 5
    static let kindJson: Int32 = 6

    var stringValue: String? {
        if case .string(let s) = self { return s }
        return nil
    }

    var boolValue: Bool? {
        if case .bool(let b) = self { return b }
        return nil
    }
}

// MARK: - TugbankClient

/// Native SQLite client for the tugbank defaults store.
///
/// Opens ~/.tugbank.db directly via the system libsqlite3. No Rust FFI,
/// no external dependencies. Reads and writes use the same schema as
/// tugbank-core (domains/entries tables, value_kind discriminator).
///
/// After a write, sends a datagram to the tugcast notification socket
/// so other processes can react.
final class TugbankClient {

    /// Shared singleton. Call `configure(path:)` once before accessing.
    private(set) static var shared: TugbankClient?

    /// Configure the shared instance. Returns the client, or nil on failure.
    @discardableResult
    static func configure(path: String) -> TugbankClient? {
        guard shared == nil else { return shared }
        guard let client = TugbankClient(path: path) else { return nil }
        shared = client
        return client
    }

    // MARK: Private state

    private var db: OpaquePointer?

    // MARK: Init / Deinit

    private init?(path: String) {
        let flags = SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE | SQLITE_OPEN_FULLMUTEX
        guard sqlite3_open_v2(path, &db, flags, nil) == SQLITE_OK else {
            if db != nil { sqlite3_close(db) }
            return nil
        }
        // Apply pragmas (must match tugbank-core/src/schema.rs)
        sqlite3_exec(db, "PRAGMA journal_mode = WAL", nil, nil, nil)
        sqlite3_exec(db, "PRAGMA foreign_keys = ON", nil, nil, nil)
        sqlite3_exec(db, "PRAGMA busy_timeout = 5000", nil, nil, nil)
        sqlite3_exec(db, "PRAGMA synchronous = NORMAL", nil, nil, nil)
    }

    deinit {
        if db != nil { sqlite3_close(db) }
    }

    // MARK: Public API

    /// Read a single value from a domain.
    func get(domain: String, key: String) -> TugbankValue? {
        let sql = "SELECT value_kind, value_i64, value_f64, value_text FROM entries WHERE domain = ? AND key = ?"
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else { return nil }
        defer { sqlite3_finalize(stmt) }

        sqlite3_bind_text(stmt, 1, (domain as NSString).utf8String, -1, nil)
        sqlite3_bind_text(stmt, 2, (key as NSString).utf8String, -1, nil)

        guard sqlite3_step(stmt) == SQLITE_ROW else { return nil }
        return decodeRow(stmt)
    }

    /// Write a value to a domain. Sends a notification datagram after success.
    @discardableResult
    func set(domain: String, key: String, value: TugbankValue) -> Bool {
        let now = ISO8601DateFormatter().string(from: Date())

        // BEGIN IMMEDIATE
        guard sqlite3_exec(db, "BEGIN IMMEDIATE", nil, nil, nil) == SQLITE_OK else { return false }

        // Ensure domain row exists
        let domainSQL = "INSERT OR IGNORE INTO domains (name, generation, updated_at) VALUES (?, 0, ?)"
        if !execBind(domainSQL, bindings: { stmt in
            sqlite3_bind_text(stmt, 1, (domain as NSString).utf8String, -1, nil)
            sqlite3_bind_text(stmt, 2, (now as NSString).utf8String, -1, nil)
        }) {
            sqlite3_exec(db, "ROLLBACK", nil, nil, nil)
            return false
        }

        // Upsert the entry
        let upsertSQL = """
            INSERT INTO entries (domain, key, value_kind, value_i64, value_f64, value_text, value_blob, updated_at)
            VALUES (?, ?, ?, ?, ?, ?, NULL, ?)
            ON CONFLICT(domain, key) DO UPDATE SET
              value_kind = excluded.value_kind, value_i64 = excluded.value_i64,
              value_f64 = excluded.value_f64, value_text = excluded.value_text,
              value_blob = excluded.value_blob, updated_at = excluded.updated_at
            """
        let (kind, i64v, f64v, textv) = encodeValue(value)
        if !execBind(upsertSQL, bindings: { stmt in
            sqlite3_bind_text(stmt, 1, (domain as NSString).utf8String, -1, nil)
            sqlite3_bind_text(stmt, 2, (key as NSString).utf8String, -1, nil)
            sqlite3_bind_int(stmt, 3, kind)
            if let v = i64v { sqlite3_bind_int64(stmt, 4, v) } else { sqlite3_bind_null(stmt, 4) }
            if let v = f64v { sqlite3_bind_double(stmt, 5, v) } else { sqlite3_bind_null(stmt, 5) }
            if let v = textv { sqlite3_bind_text(stmt, 6, (v as NSString).utf8String, -1, nil) } else { sqlite3_bind_null(stmt, 6) }
            sqlite3_bind_text(stmt, 7, (now as NSString).utf8String, -1, nil)
        }) {
            sqlite3_exec(db, "ROLLBACK", nil, nil, nil)
            return false
        }

        // Bump generation
        let genSQL = "UPDATE domains SET generation = generation + 1, updated_at = ? WHERE name = ?"
        if !execBind(genSQL, bindings: { stmt in
            sqlite3_bind_text(stmt, 1, (now as NSString).utf8String, -1, nil)
            sqlite3_bind_text(stmt, 2, (domain as NSString).utf8String, -1, nil)
        }) {
            sqlite3_exec(db, "ROLLBACK", nil, nil, nil)
            return false
        }

        // COMMIT
        guard sqlite3_exec(db, "COMMIT", nil, nil, nil) == SQLITE_OK else {
            sqlite3_exec(db, "ROLLBACK", nil, nil, nil)
            return false
        }

        // Notify tugcast
        broadcastDomainChanged(domain)
        return true
    }

    /// List all domain names.
    func listDomains() -> [String] {
        let sql = "SELECT name FROM domains ORDER BY name"
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else { return [] }
        defer { sqlite3_finalize(stmt) }

        var domains: [String] = []
        while sqlite3_step(stmt) == SQLITE_ROW {
            if let cstr = sqlite3_column_text(stmt, 0) {
                domains.append(String(cString: cstr))
            }
        }
        return domains
    }

    // MARK: Convenience

    func getString(domain: String, key: String) -> String? {
        return get(domain: domain, key: key)?.stringValue
    }

    func getBool(domain: String, key: String) -> Bool {
        guard let value = get(domain: domain, key: key) else { return false }
        switch value {
        case .bool(let b): return b
        case .string(let s): return s.caseInsensitiveCompare("true") == .orderedSame
        default: return false
        }
    }

    @discardableResult
    func setString(domain: String, key: String, value: String) -> Bool {
        return set(domain: domain, key: key, value: .string(value))
    }

    @discardableResult
    func setBool(domain: String, key: String, value: Bool) -> Bool {
        return set(domain: domain, key: key, value: .bool(value))
    }

    // MARK: Private — value encoding/decoding

    private func encodeValue(_ value: TugbankValue) -> (Int32, Int64?, Double?, String?) {
        switch value {
        case .null:
            return (TugbankValue.kindNull, nil, nil, nil)
        case .bool(let b):
            return (TugbankValue.kindBool, b ? 1 : 0, nil, nil)
        case .i64(let n):
            return (TugbankValue.kindI64, n, nil, nil)
        case .f64(let f):
            return (TugbankValue.kindF64, nil, f, nil)
        case .string(let s):
            return (TugbankValue.kindString, nil, nil, s)
        case .json(let obj):
            if let data = try? JSONSerialization.data(withJSONObject: obj),
               let text = String(data: data, encoding: .utf8) {
                return (TugbankValue.kindJson, nil, nil, text)
            }
            return (TugbankValue.kindNull, nil, nil, nil)
        }
    }

    private func decodeRow(_ stmt: OpaquePointer?) -> TugbankValue? {
        let kind = sqlite3_column_int(stmt, 0)
        switch kind {
        case TugbankValue.kindNull:
            return .null
        case TugbankValue.kindBool:
            return .bool(sqlite3_column_int64(stmt, 1) != 0)
        case TugbankValue.kindI64:
            return .i64(sqlite3_column_int64(stmt, 1))
        case TugbankValue.kindF64:
            return .f64(sqlite3_column_double(stmt, 2))
        case TugbankValue.kindString:
            if let cstr = sqlite3_column_text(stmt, 3) {
                return .string(String(cString: cstr))
            }
            return .string("")
        case TugbankValue.kindJson:
            if let cstr = sqlite3_column_text(stmt, 3) {
                let text = String(cString: cstr)
                if let data = text.data(using: .utf8),
                   let obj = try? JSONSerialization.jsonObject(with: data) {
                    return .json(obj)
                }
            }
            return .null
        default:
            return nil
        }
    }

    // MARK: Private — SQL helpers

    private func execBind(_ sql: String, bindings: (OpaquePointer?) -> Void) -> Bool {
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else { return false }
        defer { sqlite3_finalize(stmt) }
        bindings(stmt)
        return sqlite3_step(stmt) == SQLITE_DONE
    }

    // MARK: Private — notification

    private func broadcastDomainChanged(_ domain: String) {
        let socketPath = FileManager.default.temporaryDirectory
            .appendingPathComponent("tugbank-notify.sock").path

        let fd = socket(AF_UNIX, SOCK_DGRAM, 0)
        guard fd >= 0 else { return }
        defer { close(fd) }

        var addr = sockaddr_un()
        addr.sun_family = sa_family_t(AF_UNIX)
        let pathBytes = socketPath.utf8CString
        guard pathBytes.count <= MemoryLayout.size(ofValue: addr.sun_path) else { return }
        _ = withUnsafeMutablePointer(to: &addr.sun_path) { sunPath in
            pathBytes.withUnsafeBufferPointer { buf in
                memcpy(sunPath, buf.baseAddress!, buf.count)
            }
        }

        let domainBytes = Array(domain.utf8)
        withUnsafePointer(to: &addr) { addrPtr in
            addrPtr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sockaddrPtr in
                _ = sendto(fd, domainBytes, domainBytes.count, 0, sockaddrPtr,
                          socklen_t(MemoryLayout<sockaddr_un>.size))
            }
        }
    }
}
